File: 97-interleaving-string/interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        n = len(s1)
        m = len(s2)
        t = len(s3)
        if n + m != t:
            return False
        @cache
        def recur(i: int, j:int):
            if i == -1 and j == -1:
                return True
            canForm = False
            canForm = canForm or (i != -1 and s1[i] == s3[i+j +1] and recur(i-1, j))
            canForm = canForm or (j != -1 and s2[j] == s3[i+j +1] and recur(i, j-1))
            return canForm

        return recur(n-1, m-1)

File: 97-interleaving-string/interleaving-string.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int n = s1.size();
        int m = s2.size();
        int k = s3.size();
        if(n+m != k) return false;
        vector<vector<bool>> dp(n+1, vector<bool>(m+1, false));
        dp[0][0] = true;

        for(int i = 1; i < n+1; i++){
            if(s1[i-1] == s3[i-1] && dp[i-1][0]) dp[i][0] = true;
        }
        for(int i = 1; i < m+1; i++){
            if(s2[i-1] == s3[i-1] && dp[0][i-1]) dp[0][i] = true;
        }

        for(int i = 1; i < n+1; i++){
            for(int j = 1; j < m+1; j++){
                if(dp[i-1][j] && s1[i-1] == s3[i+j-1]) dp[i][j] = true;
                if(dp[i][j-1] && s2[j-1] == s3[i+j-1]) dp[i][j] = true;
            }
        }
        return dp[n][m];
    }
};